parser/src/lexer.rs

```diff
@@ -644,6 +696,12 @@ where
                     spaces = 0;
                     tabs = 0;
                 }
+                // https://github.com/ipython/ipython/blob/635815e8f1ded5b764d66cacc80bbe25e9e2587f/IPython/core/inputtransformer2.py#L345
+                Some('%' | '!' | '?' | '/' | ';' | ',') if self.mode == Mode::Jupyter => {
+                    self.lex_and_emit_magic_command()?;
+                    spaces = 0;
+                    tabs = 0;
+                }
```

Comment:
Does this result in an endless loop if it turns out that it isn't a magic command after all (because it doesn't eat any characters)

Reply:
What are you considering in "that isn't a magic command"? Do you mean that if the token is followed directly by a newline (empty command string)? If so, then no it'll still eat the escape characters.

For example, if a file contains only the escape tokens:
```
%
%%
!
!!
?
??
/
,
;
```

The following tokens will be produced:
```
0 MagicCommand {
    value: "",
    kind: Magic,
} 1
1 NonLogicalNewline 2
2 MagicCommand {
    value: "",
    kind: Magic2,
} 4
4 NonLogicalNewline 5
5 MagicCommand {
    value: "",
    kind: Shell,
} 6
6 NonLogicalNewline 7
7 MagicCommand {
    value: "",
    kind: ShCap,
} 9
9 NonLogicalNewline 10
10 MagicCommand {
    value: "",
    kind: Help,
} 11
11 NonLogicalNewline 12
12 MagicCommand {
    value: "",
    kind: Help2,
} 14
14 NonLogicalNewline 15
15 MagicCommand {
    value: "",
    kind: Paren,
} 16
16 NonLogicalNewline 17
17 MagicCommand {
    value: "",
    kind: Quote,
} 18
18 NonLogicalNewline 19
19 MagicCommand {
    value: "",
    kind: Quote2,
} 20
20 NonLogicalNewline 21
```

Reply:
That makes sense if only the start characters are required for a magic command.